lte/gateway/c/connection_tracker/EventTracker.cpp

```diff
@@ -0,0 +1,255 @@
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "EventTracker.h"
+
+static int parse_ip_cb(const struct nlattr* attr, void* data) {
+  const struct nlattr** tb = (const struct nlattr**) data;
+  int type                 = mnl_attr_get_type(attr);
+
+  if (mnl_attr_type_valid(attr, CTA_IP_MAX) < 0) return MNL_CB_OK;
+
+  switch (type) {
+    case CTA_IP_V4_SRC:
+    case CTA_IP_V4_DST:
+      if (mnl_attr_validate(attr, MNL_TYPE_U32) < 0) {
+        perror("mnl_attr_validate");
+        return MNL_CB_ERROR;
+      }
+      break;
+  }
+  tb[type] = attr;
+  return MNL_CB_OK;
+}
+
+static void parse_ip(const struct nlattr* nest, struct flow_information* flow) {
+  struct nlattr* tb[CTA_IP_MAX + 1] = {};
+
+  mnl_attr_parse_nested(nest, parse_ip_cb, tb);
+  if (tb[CTA_IP_V4_SRC]) {
+    struct in_addr* in =
+        (struct in_addr*) mnl_attr_get_payload(tb[CTA_IP_V4_SRC]);
+    // printf("src=%s ", inet_ntoa(*in));
+    flow->saddr = in->s_addr;
+  }
+  if (tb[CTA_IP_V4_DST]) {
+    struct in_addr* in =
+        (struct in_addr*) mnl_attr_get_payload(tb[CTA_IP_V4_DST]);
+    // printf("dst=%s ", inet_ntoa(*in));
+    flow->daddr = in->s_addr;
+  }
+}
+
+static int parse_proto_cb(const struct nlattr* attr, void* data) {
+  const struct nlattr** tb = (const struct nlattr**) data;
+  int type                 = mnl_attr_get_type(attr);
+
+  if (mnl_attr_type_valid(attr, CTA_PROTO_MAX) < 0) return MNL_CB_OK;
+
+  switch (type) {
+    case CTA_PROTO_NUM:
+    case CTA_PROTO_ICMP_TYPE:
+    case CTA_PROTO_ICMP_CODE:
+      if (mnl_attr_validate(attr, MNL_TYPE_U8) < 0) {
+        perror("mnl_attr_validate");
+        return MNL_CB_ERROR;
+      }
+      break;
+    case CTA_PROTO_SRC_PORT:
+    case CTA_PROTO_DST_PORT:
+    case CTA_PROTO_ICMP_ID:
+      if (mnl_attr_validate(attr, MNL_TYPE_U16) < 0) {
+        perror("mnl_attr_validate");
+        return MNL_CB_ERROR;
+      }
+      break;
+  }
+  tb[type] = attr;
+  return MNL_CB_OK;
+}
+
+static void parse_proto(
+    const struct nlattr* nest, struct flow_information* flow) {
+  struct nlattr* tb[CTA_PROTO_MAX + 1] = {};
+
+  mnl_attr_parse_nested(nest, parse_proto_cb, tb);
+  if (tb[CTA_PROTO_NUM]) {
+    // printf("proto=%u ", mnl_attr_get_u8(tb[CTA_PROTO_NUM]));
+    flow->l4_proto = mnl_attr_get_u8(tb[CTA_PROTO_NUM]);
+  }
+  if (tb[CTA_PROTO_SRC_PORT]) {
+    // printf("sport=%u ", ntohs(mnl_attr_get_u16(tb[CTA_PROTO_SRC_PORT])));
+    flow->sport = mnl_attr_get_u8(tb[CTA_PROTO_NUM]);
+  }
+  if (tb[CTA_PROTO_DST_PORT]) {
+    // printf("dport=%u ", ntohs(mnl_attr_get_u16(tb[CTA_PROTO_DST_PORT])));
```

Comment:
This might be useful for debugging, can you use logging api for this.

Reply:
I think printing the packet is cleaner than printing these line by line (its printed completely a bit later.